### ISSUE TITLE
SHACL generator: support LinkML rules

### ISSUE THREAD

**Issue by elichad** [linkml/linkml#2464]
**What is your feature request?**
The SHACL generator should support LinkML [rules](https://linkml.io/linkml/schemas/advanced.html#rules). Currently they are not considered.

**How important is this feature to you?** Select from the options below:
• Medium - can do work without it; but it's important (e.g. to save time or for convenience)

**Additional context**
I have yet to determine if workarounds are available for my specific use case, or if this is absolutely blocking for me.

**Comment by mih** [comment]
Possibly related: Slots' `ifabsent` is also not supported by the SHACL generator and raises `NotImplemented`.

**Comment by elichad** [comment]
@mih that sounds like a separate issue to me - there are many LinkML features not supported by the SHACL generator currently, so I think it's better to track them separately. 

Also, yours sounds like an easier fix than mine :) There's already some infrastructure for converting `ifabsent`, it might just be missing support for whatever datatype you're using: linkml/generators/shacl/shacl_ifabsent_processor.py

(I'm not a LinkML dev but I've been poking around the SHACL generator code myself)

**Comment by jsheunis** [comment]
> there are many LinkML features not supported by the SHACL generator currently, so I think it's better to track them separately

> There's already some infrastructure for converting ifabsent, it might just be missing support for whatever datatype you're using

Agreed. I created an issue here: https://github.com/linkml/linkml/issues/2502

**Comment by jsheunis** [comment]
Coming back to this issue to up it again. We're working with tools to generate forms from SHACL shapes, and so-called conditionals are essential for an intuitive UX in such forms, e.g. if a user selects Option A in Question 1, then Questions 8-10 should be mandatory, otherwise not. Another quite simple case is to support something similar to LinkML's string serialization and ifabsent processors, but with those same conditionals exported to SHACL (e.g. if the user does not complete field X, generate its value by concatenating the values of fields Y and Z).

From my limited search, it seems that LinkML would need to support [SHACL advance features](https://www.w3.org/TR/shacl-af) to some extent in order to allow the export of SHACL that include expressions, functions, and/or rules. Quoting from these docs:

> SHACL functions can also be used declaratively in frameworks such as the SHACL [node expressions](https://www.w3.org/TR/shacl-af/#dfn-node-expressions) which are used in [SHACL rules](https://www.w3.org/TR/shacl-af/#dfn-shacl-rules). In those scenarios they may be used to perform data transformations such as string concatenation.

So, e.g., string serialization could be codified for a given use case as follows:

```
[  ex:concat ( 
		[ sh:path ex:firstName ]
		[ sh:path ex:lastName ] 
	)
] .
```

There are of course many forms of logic that one might want to codify.

For myself, the next step will be to investigate SHACL advanced features (and any other SHACL features that could support the above need) more thoroughly and put together a real-world use case to work towards.

Hoping there are others out there with similar interests so that we can crowdsource this process.

**Comment by jsheunis** [comment]
An update:

I went on an investigative expedition into [SHACL Core Constraint Components](https://www.w3.org/TR/shacl/#core-components) and [SHACL Advanced Features 1.1](https://w3c.github.io/shacl/shacl-af/) (see here: https://github.com/psychoinformatics-de/shacl-vue/issues/204, or perhaps don't if you are averse to reading many words).

From my POV, the latest states of these specifications are generally expressive enough to be able to represent LinkML `rules` (as well as `equals_expression` and `string_serialization` statements) in SHACL, using a mixture of core constraint components, and SHACL node expressions, functions, expression constraints, and rules.

There are a number of practical challenges to address though, IMO:

### 1. Constraints vs data inference

Some rule pre-/postconditions might add an additional constraint (e.g. if the value of `slotA` is `X`, then `slotB` should be `required`) while others rather imply inference (e.g. if the value of `age` is larger than 50, then the value of `over_fifty` should be `true`).

An interpreted constraint for which a direct mapping to a SHACL Core Constraint Component does not exist would then map to a SHACL expression constraint `sh:expression` (strictly speaking), while a rule for inference would map to an `sh:rule` (likely `sh:TripleRule` to add new data)

Because of this it might be necessary to inform the `gen-shacl` process as to what the goal of the SHACL generation is, i.e. constraints / inference / both (similar to the `--infer` option of `linkml-convert`). What would complicate this is if different schema components might have different goals.

### 2. The non-deterministic way in which LinkML schema components map to SHACL

The constraint vs inference problem is a special case of this, given that the same LinkML schema component could map to either `sh:expression` or `sh:rule`. However, there is also the larger general challenge of deciding how to express the relevant LinkML rules into SHACL in the absence of a deterministic mapping.

Consider the example of a LinkML class with slots `skos:prefLabel` (required) and `ex:altLabel` (not required), where there is a rule that the `ex:altLabel` should be different from the `skos:prefLabel`:

```yaml
classes:
  MyClass:
    slots:
      - prefLabel
      - altLabel
    slot_usage:
      prefLabel:
        required: true
    rules:
      - postcondiotions:
          slot_conditions:
            altLabel:
              none_of:
                equals_expression: "{prefLabel}"
```

_(Note that I did not include `preconditions` since I am unsure how to express the condition that `alt_label` should exist for the rule to apply; although I am also unsure whether this is specifically necessary in this example)._

Since this is a constraint, and there is a known [`sh:disjoint` property pair constraint](https://www.w3.org/TR/shacl/#DisjointConstraintComponent) in SHACL Clore, the exported SHACL could be fairly straightforward:

```ttl
ex:MyClassShape a sh:NodeShape ;
    sh:property [ sh:datatype xsd:string ;
            sh:nodeKind sh:Literal ;
            sh:order 0 ;
            sh:minCount 1 ;
            sh:maxCount 1 ;
            sh:path skos:prefLabel ],
        [ sh:datatype xsd:string ;
            sh:nodeKind sh:Literal ;
            sh:order 1 ;
            sh:path ex:altLabel ;
            sh:disjoint skos:prefLabel ],
    sh:targetClass ex:MyClass .
```

However, the constraint could also be expressed with `sh:expression` and a combination of a custom defined function `ex:notEqual` and path expressions as arguments:

```ttl
ex:MyClassShape a sh:NodeShape ;
    sh:property [ sh:datatype xsd:string ;
            sh:nodeKind sh:Literal ;
            sh:order 0 ;
            sh:minCount 1 ;
            sh:maxCount 1 ;
            sh:path skos:prefLabel ],
        [ sh:datatype xsd:string ;
            sh:nodeKind sh:Literal ;
            sh:order 1 ;
            sh:path ex:altLabel ],
    sh:targetClass ex:MyClass ;
    sh:expression [
        ex:notEqual ( [ sh:path skos:prefLabel] [ sh:path ex:altLabel ] );
    ] .
```

There are likely numerous other ways to express the same constraint as well; and similarly various ways to express rules. One should note he SHACL CORE Constraint Components and SHACL AF node expressions, while there are many, do not necessarily cover (deterministically) all possibilities of the LinkML, which could well mean that SHACL functions will be required. This would broaden the scope of possible mappings and complicate the non-determinism challenge even more.

### 3. Insufficient (or inaccessible) information regarding implementation details

_Caveat: this could be the result of my inexperience with the code base._

In terms of LinkML implementation, the code that `gen-shacl` might be able to reuse or hook into seems to be the utility functionality in `linkml-runtime`, specifically:
- [eval_utils](https://github.com/linkml/linkml-runtime/blob/main/linkml_runtime/utils/eval_utils.py)
- [rule_utils](https://github.com/linkml/linkml-runtime/blob/main/linkml_runtime/utils/ruleutils.py)
- [inference_utils](https://github.com/linkml/linkml-runtime/blob/main/linkml_runtime/utils/inference_utils.py)

I was (yet) unable to determine how exactly these utilities structure the expressions and rules derived from a given schema, and whether those structured expressions and rules could somehow be mapped in a practical way to SHACL.

Improved documentation and examples regarding LinkML rules could be useful here.
